src/Aspire.Dashboard/Telemetry/DashboardTelemetrySender.cs

```diff
@@ -0,0 +1,202 @@
+                    {
+                        var (context, requestFunc) = operation;
+                        try
+                        {
+                            await requestFunc(_httpClient, GetResponseProperty).ConfigureAwait(false);
+
+                            // Double check properties are set.
+                            foreach (var property in context.Properties)
+                            {
+                                if (!property.HasValue)
+                                {
+                                    _logger.LogWarning("Unset context property.");
+                                }
+                            }
+                        }
+                        catch (Exception ex)
+                        {
+                            _logger.LogWarning(ex, "Failed to send telemetry request.");
+                        }
+                    }
+                }
+            }
+            finally
+            {
+                _logger.LogInformation("Ending sender loop.");
+            }
+        });
+    }
+
+    private HttpClient CreateHttpClient(Uri debugSessionUri, string token, X509Certificate2 cert)
+    {
+        var handler = new HttpClientHandler
+        {
+            ServerCertificateCustomValidationCallback = (_, c, _, e) =>
+            {
+                // Server certificate is already considered valid.
+                if (e == SslPolicyErrors.None)
+                {
+                    return true;
+                }
+
+                // Certificate isn't immediately valid. Check if it is the same as the one we expect.
+                return string.Equals(
+                    cert.GetCertHashString(HashAlgorithmName.SHA256),
+                    c?.GetCertHashString(HashAlgorithmName.SHA256),
+                    StringComparison.OrdinalIgnoreCase);
+            }
+        };
+        var resolvedHandler = CreateHandler?.Invoke(handler) ?? handler;
+        var client = new HttpClient(resolvedHandler)
+        {
+            BaseAddress = debugSessionUri,
+            DefaultRequestHeaders =
+            {
+                { "Authorization", $"Bearer {token}" },
+                { "User-Agent", "Aspire Dashboard" }
+            }
+        };
+
+        return client;
+    }
+
+    private object GetResponseProperty(OperationContextProperty propertyId)
+    {
+        return propertyId.GetValue();
+    }
+
+    public async Task<bool> TryStartTelemetrySessionAsync()
+    {
+        Debug.Assert(_isEnabled is null, "Telemetry session has already been started.");
+
+        _isEnabled = await TryStartTelemetrySessionCoreAsync().ConfigureAwait(false);
+        if (_isEnabled.Value)
+        {
+            StartSendLoop();
+        }
+
+        return _isEnabled.Value;
+    }
+
+    private async Task<bool> TryStartTelemetrySessionCoreAsync()
+    {
+        if (HasDebugSession(_options.Value.DebugSession, out var debugSessionUri, out var token, out var certData))
```

Comment:
cert data is optional - https://github.com/dotnet/aspire/blob/main/docs/specs/IDE-execution.md#enabling-ide-execution